### PR TITLE
types: rewrite API types

### LINKED ISSUE #267: Rewrite API types
Currently, API types are mixed with the library types and some of the types are erroneous and somewhat inaccurate; a complete rewrite of API types would be great!

### PR DESCRIPTION
**Pull request by ayntee**
Closes #267

### CODE REVIEW

File: src/api/controllers/cache.ts [hunk 1]
```diff
@@ -110,28 +111,34 @@ function filter(
...
 async function filter(
   table: TableName,
+  // deno-lint-ignore no-explicit-any
   callback: (value: any, key: string) => boolean,
 ) {
   return cache[table].filter(callback);
 }
 
-export let cacheHandlers = {
+export const cacheHandlers = {
```
Comment (ayntee) [inline src/api/controllers/cache.ts:123]: `const` -> `let`

Comment (ayntee) [inline src/api/controllers/cache.ts:123]:
```suggestion
// deno-lint-ignore prefer-const
export let cacheHandlers = {
```

File: src/api/controllers/interactions.ts [hunk 2]
```diff
@@ -1,27 +1,35 @@
...
+import { camelKeysToSnakeCase } from "../../util/utils.ts";
 import { structures } from "../structures/mod.ts";
 
-export async function handleInternalInteractionCreate(data: DiscordPayload) {
+export async function handleInternalInteractionCreate(data: GatewayPayload) {
   if (data.t !== "INTERACTION_CREATE") return;
 
-  const payload = data.d as InteractionCommandPayload;
+  const payload = data.d as Interaction;
+  const member: GuildMemberPayload | undefined = camelKeysToSnakeCase(
```
Comment (ayntee) [inline src/api/controllers/interactions.ts:15]: @itohatweb can you remove this? No breaking changes.

File: src/api/controllers/cache.ts (hunk 1 above)
Comment (itohatweb) [reply src/api/controllers/cache.ts:123]: lgtm

**Comment by itohatweb** [comment]
Found something that needs to be done. I'm already on that, please don't merge.

File: src/api/controllers/interactions.ts [hunk 3]
```diff
@@ -1,27 +1,35 @@
...
+    (await structures.createMember(payload.member, payload.guild_id))
+      .guildMember(payload.guild_id),
+  );
+  if (!member) return;
```
[the remark this answers is not in the document]
Comment (itohatweb) [reply src/api/controllers/interactions.ts]: Well thats needed. Any Idea to change?

File: src/api/handlers/channel.ts [hunk 4]
```diff
@@ -214,27 +224,20 @@ export async function sendMessage(
...
-          replied_user: content.mentions.repliedUser,
-        }
-        : undefined,
-      message_reference: {
-        message_id: content.replyMessageID,
-      },
+      ...camelKeysToSnakeCase(content),
+      message_reference: typeof content.reply === "string"
+        ? { message_id: content.reply }
+        : camelKeysToSnakeCase(content.reply),
```
[the remark this answers is not in the document]
Comment (itohatweb) [reply src/api/handlers/channel.ts:230]: What do you want here?

File: src/api/controllers/messages.ts [hunk 5]
```diff
@@ -22,16 +22,31 @@ export async function handleInternalMessageCreate(data: DiscordPayload) {
...
       structures.createMember(
-        { ...mention.member, user: mention },
+        {
+          nick: mention.nick === undefined ? null : mention.nick,
+          roles: mention.roles || [],
+          joined_at: mention.joined_at || "",
+          deaf: mention.deaf || false,
+          mute: mention.mute || false,
+          user: mention,
+        },
```
[the remark this answers is not in the document]
Comment (itohatweb) [reply src/api/controllers/messages.ts:50]: Typescript does not like if I don't put default values here. I try to search a better solution.

File: src/api/handlers/channel.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (itohatweb) [reply src/api/handlers/channel.ts:230]: Actually I made this change to prevent breaking changes

File: src/api/handlers/channel.ts [hunk 6]
```diff
@@ -289,7 +291,14 @@ export async function createInvite(
...
   }
-  return RequestManager.post(endpoints.CHANNEL_INVITES(channelID), options);
+
+  return RequestManager.post(endpoints.CHANNEL_INVITES(channelID), {
+    ...options,
+    max_age: options.maxAge,
+    max_uses: options.maxUses,
+    target_user: options.targetUser,
+    target_user_type: options.targetUserType,
+  });
```
Comment (itohatweb) [inline src/api/handlers/channel.ts:301]: Use the converter function

File: src/api/handlers/guild.ts [hunk 7]
```diff
@@ -247,45 +304,45 @@ export async function createEmoji(
...
+      options,
+    ),
+  ) as Emoji;
 }
 
 /** Delete the given emoji. Requires the MANAGE_EMOJIS permission. Returns 204 No Content on success. */
 export async function deleteEmoji(
   guildID: string,
-  id: string,
-  reason?: string,
```
Comment (itohatweb) [inline src/api/handlers/guild.ts:278]: Add this back

File: src/api/handlers/guild.ts [hunk 8]
```diff
@@ -296,8 +353,7 @@ export function emojiURL(id: string, animated = false) {
 /** Create a new role for the guild. Requires the MANAGE_ROLES permission. */
 export async function createGuildRole(
   guildID: string,
-  options: CreateRoleOptions,
-  reason?: string,
```
Comment (itohatweb) [inline src/api/handlers/guild.ts:300]: Add this back

File: src/api/handlers/guild.ts [hunk 9]
```diff
@@ -571,7 +676,7 @@ export async function ban(guildID: string, id: string, options: BanOptions) {
 
   return RequestManager.put(
     endpoints.GUILD_BAN(guildID, id),
-    { ...options, delete_message_days: options.days },
+    { ...snakeKeysToCamelCase(options) },
```
Comment (itohatweb) [inline src/api/handlers/guild.ts:679]: Spreading is unnecessary

File: src/api/handlers/member.ts [hunk 10]
```diff
@@ -53,7 +60,6 @@ export async function addRole(
   guildID: string,
   memberID: string,
   roleID: string,
-  reason?: string,
```
Comment (itohatweb) [inline src/api/handlers/member.ts:56]: Add this back

File: src/api/handlers/member.ts [hunk 11]
```diff
@@ -86,7 +91,6 @@ export async function removeRole(
...
```
Comment (itohatweb) [inline src/api/handlers/member.ts:89]: Add this back

File: src/api/handlers/member.ts [hunk 12]
```diff
@@ -142,7 +145,7 @@ export async function sendDirectMessage(
 }
 
 /** Kick a member from the server */
-export async function kick(guildID: string, memberID: string, reason?: string) {
```
Comment (itohatweb) [inline src/api/handlers/member.ts:145]: Add this back

File: src/api/handlers/message.ts [hunk 13]
```diff
@@ -33,7 +29,6 @@ export async function deleteMessageByID(
 /** Delete a message */
 export async function deleteMessage(
   message: Message,
-  reason?: string,
```
Comment (itohatweb) [inline src/api/handlers/message.ts:36]: Add this back

File: src/api/handlers/webhook.ts [hunk 14]
```diff
@@ -309,10 +357,10 @@ export function deleteSlashResponse(
 /** To edit your response to a slash command. If a messageID is not provided it will default to editing the original response. */
 export function editSlashResponse(
   token: string,
-  options: EditSlashResponseOptions,
+  options: EditWebhookMessageOptions,
 ) {
   return RequestManager.patch(
     endpoints.INTERACTION_ORIGINAL_ID_TOKEN(botID, token),
-    options,
+    { ...camelKeysToSnakeCase(options) },
```
Comment (itohatweb) [inline src/api/handlers/webhook.ts:364]: Spreading is unnecessary

File: src/api/structures/guild.ts [hunk 15]
```diff
@@ -69,10 +71,10 @@ const baseGuild: Partial<Guild> = {
     return cache.members.get(this.ownerID!);
   },
   get partnered() {
-    return Boolean(this.features?.includes("PARTNERED"));
+    return Boolean(this.features?.includes(GuildFeatures.PARTNERED));
   },
   get verified() {
-    return Boolean(this.features?.includes("VERIFIED"));
+    return Boolean(this.features?.includes(GuildFeatures.VERIFIED));
```
Comment (itohatweb) [inline src/api/structures/guild.ts:77]: We want strings back

File: src/rest/request_manager.ts [hunk 16]
```diff
@@ -310,7 +311,7 @@ function runMethod(
 async function logErrors(response: Response, errorStack?: unknown) {
   try {
     const error = await response.json();
-    console.error(error);
+    console.error(Deno.inspect(error, { depth: 10 }));
```
Comment (itohatweb) [inline src/rest/request_manager.ts:314]: remove this

**Review by itohatweb — CHANGES_REQUESTED** [review]

**Comment by itohatweb** [comment]
TODO: fix interaction module

**Pull request closed without merging**